text/0000-reactive-assignments.md

```diff
@@ -0,0 +1,944 @@
+
+> (The 🐃 emoji used throughout this document indicates a yak that needs shaving.)
+
+Here, we have declared a single state variable, `count`, with an initial value of `0`. The code inside the `<script>` block runs **once per component instance, when the instance is created**.
+
+When clicking the button, the text inside the button should update to reflect the current value. This can be achieved by scheduling an update after every assignment expression (these are trivial to discover via AST traversal), resulting in something resembling this pseudo-code:
+
+```js
+button.addEventListener('click', () => {
+  count += 1;
+  __update({ count: true });
+});
+```
+
+`__update` is a framework-provided function that causes the view to update, with knowledge of which values have changed. The implementation of `__update` is unimportant for now, other than to note that it schedules an update **asynchronously** in contrast to Svelte's current model, for reasons to be discussed later.
+
+This code transformation also applies to assignments inside the `<script>` block. This component...
+
+```html
+<script>
+  let count = 0;
+  const incr = () => {
+    count += 1;
+  };
+</script>
+
+<button on:click="incr()">
+  clicks: {count}
+</button>
+```
+
+...would be transformed to something like this:
+
+```js
+function create_main_fragment(component, ctx) {
+  // ... the code that creates and maintains the view —
+  // this will be relatively unaffected by this proposal
+}
+
+const Component = defineComponent((__update) => {
+  let count = 0;
+  const incr = () => {
+    count += 1;
+    __update({ count: true });
+  };
+
+  // this creates the top-level `ctx` variable for `create_main_fragment`
+  return () => { count };
+}, create_main_fragment);
+```
+
+**This behaviour might seem surprising, even shocking at first.** Variable assignment does not have side-effects in JavaScript, meaning that this is arguably something else — [SvelteScript](https://mobile.twitter.com/youyuxi/status/1057291776724271104), perhaps. In practice though, developers readily embrace 'magic' that makes their day-to-day lives easier as long as the mechanisms are ultimately easy to understand — something observed with React Hooks, Vue's reactivity system, Immer's approach to immutable data, and countless other examples. This does underscore the need for this code transformation to be well-documented and explained, however.
+
+
+### Props
+
+Many frameworks (and also web components) have a conceptual distinction between 'props' (values passed *to* a component) and 'state' (values that are internal to the component). Svelte does not. This is a shortcoming.
+
+The proposed solution is to use the `export` keyword to declare a value that can be set from outside as a prop:
```

Comment:
One piece of feedback I've received objected to using `export` because 'props flow in, not out' — this feedback was accompanied by an analogy I won't repeat here.

It's not strictly accurate, since this RFC proposes that properties be attached to the component instances as accessors, as happens with web components, and it is a means by which we can specify a custom component API. Nevertheless, people are likely to react similarly, so we should perhaps adjust the language around this to discuss alternatives and explain why they're unsuitable. (The suggestion here was to use a magic `import`, but aside from the inherent problems with magic imports, that wouldn't allow us to specify default values.)

Reply:
Updated the text, resolving this item